crates/aptos/src/move_tool/mod.rs

```diff
@@ -1223,130 +1214,134 @@ impl Display for FunctionArgType {
             FunctionArgType::U128 => bcs::to_bytes(
                 &u128::from_str(arg)
                     .map_err(|err| CliError::UnableToParse("u128", err.to_string()))?,
-            ),
+            )
+            .map_err(|err| CliError::BCS("arg", err)),
             FunctionArgType::U256 => bcs::to_bytes(
                 &U256::from_str(arg)
                     .map_err(|err| CliError::UnableToParse("u256", err.to_string()))?,
-            ),
+            )
+            .map_err(|err| CliError::BCS("arg", err)),
             FunctionArgType::Raw => {
-                let raw = hex::decode(arg)
-                    .map_err(|err| CliError::UnableToParse("raw", err.to_string()))?;
-                Ok(raw)
+                hex::decode(arg).map_err(|err| CliError::UnableToParse("raw", err.to_string()))
+            },
+        }
+    }
+
+    /// Recursively parse argument JSON into BCS representation.
+    pub fn parse_arg_json(&self, arg: &serde_json::Value) -> CliTypedResult<ArgWithType> {
+        match arg {
+            serde_json::Value::Bool(value) => Ok(ArgWithType {
+                _ty: self.clone(),
+                _vector_depth: 0,
+                arg: self.parse_arg_str(value.to_string().as_str())?,
+            }),
+            serde_json::Value::Number(value) => Ok(ArgWithType {
+                _ty: self.clone(),
+                _vector_depth: 0,
+                arg: self.parse_arg_str(value.to_string().as_str())?,
+            }),
+            serde_json::Value::String(value) => Ok(ArgWithType {
+                _ty: self.clone(),
+                _vector_depth: 0,
+                arg: self.parse_arg_str(value.as_str())?,
+            }),
+            serde_json::Value::Array(_) => {
+                let mut bcs: Vec<u8> = vec![]; // BCS representation of argument.
+                let mut common_sub_arg_depth = None;
+                // Prepend argument sequence length to BCS bytes vector.
+                write_u64_as_uleb128(&mut bcs, arg.as_array().unwrap().len());
+                // Loop over all of the vector's sub-arguments, which may also be vectors:
+                for sub_arg in arg.as_array().unwrap() {
+                    let ArgWithType {
+                        _ty: _,
+                        _vector_depth: sub_arg_depth,
+                        arg: mut sub_arg_bcs,
+                    } = self.parse_arg_json(sub_arg)?;
+                    // Verify all sub-arguments have same depth.
+                    if let Some(check_depth) = common_sub_arg_depth {
+                        if check_depth != sub_arg_depth {
+                            return Err(CliError::CommandArgumentError(
+                                "Variable vector depth".to_string(),
+                            ));
+                        }
+                    };
+                    common_sub_arg_depth = Some(sub_arg_depth);
+                    bcs.append(&mut sub_arg_bcs); // Append sub-argument BCS.
```

Comment:
My only concern is that we're reimplementing BCS, but probably points out that maybe BCS needs a more interactive streaming approach in Rust like you're doing here.

Reply:
Agreed. I go into more depth in the above comment https://github.com/aptos-labs/aptos-core/pull/7790#discussion_r1178402832